File: bms/go/go-gt/Dijkstra/Dijkstra.go
package test

import (
	"math"
)

var ()

const ()

func Dijkstra(G *Matrix, i int64) (p []int64) {
	p = make([]int64, G.N)
	h := newHeap(G.N)
	h.W[i] = 0
	h.swap(i, 0)
	for h.N > 0 {
		i = h.pop()
		if h.W[i] == math.MaxInt64 {
			return p
		}
		h.update(p, i, G)
	}
	return p
}

func newHeap(n int64) (h heap) {
	h.N = n
	h.I = make([]int64, n)
	h.A = make([]int64, n)
	h.W = make([]int64, n)
	var i int64
	for i = 0; i < n; i++ {
		h.I[i] = i
		h.A[i] = i
		h.W[i] = math.MaxInt64
	}
	return h
}

func (h heap) swap(a, b int64) {
	i, j := h.I[a], h.I[b]
	h.I[a], h.I[b] = h.I[b], h.I[a]
	h.A[i], h.A[j] = b, a
}

func (h *heap) pop() (i int64) {
	i = h.I[0]
	h.N--
	h.swap(0, h.N)
	h.down(0)
	return i
}

func (h heap) down(i int64) {
	for {
		left := 2*i + 1
		if left >= h.N {
			break
		}
		j := left
		if right := left + 1; right < h.N && !h.less(left, right) {
			j = right
		}
		if h.less(i, j) {
			break
		}
		h.swap(i, j)
		i = j
	}
}

func (h heap) less(a, b int64) bool {
	i, j := h.I[a], h.I[b]
	return h.W[i] < h.W[j]
}

func (h heap) update(p []int64, i int64, G *Matrix) {
	var j int64
	for j = 0; j < G.N; j++ {
		if G.Get(i, j) > 0 {
			if h.W[i]+G.Get(i, j) < h.W[j] {
				p[j] = i + 1
				h.W[j] = h.W[i] + G.Get(i, j)
				h.up(h.A[j])
			}
		}
	}
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

func (h heap) up(j int64) {
	for {
		i := (j - 1) / 2
		if i == j || h.less(i, j) {
			break
		}
		h.swap(i, j)
		j = i
	}
}

type Matrix struct {
	N int64
	A []int64
}

type heap struct {
	N int64
	I []int64
	A []int64
	W []int64
}

File: translations_only/fail/go/go-gt/Dijkstra/gpt4/Restart/Dijkstra.rs

use std::i64;

struct Matrix {
    n: i64,
    a: Vec<i64>,
}

struct Heap {
    n: i64,
    i: Vec<i64>,
    a: Vec<i64>,
    w: Vec<i64>,
}

fn dijkstra(g: &Matrix, i: i64) -> Vec<i64> {
    let mut p = vec![0; g.n as usize];
    let mut h = new_heap(g.n);
    h.w[i as usize] = 0;
    swap(&mut h, i, 0);
    while h.n > 0 {
        let i = pop(&mut h);
        if h.w[i as usize] == i64::MAX {
            return p;
        }
        update(&mut h, &mut p, i, g);
    }
    p
}

fn new_heap(n: i64) -> Heap {
    Heap {
        n,
        i: (0..n).collect(),
        a: (0..n).collect(),
        w: vec![i64::MAX; n as usize],
    }
}

fn swap(h: &mut Heap, a: i64, b: i64) {
    let (i, j) = (h.i[a as usize], h.i[b as usize]);
    h.i.swap(a as usize, b as usize);
    h.a.swap(i as usize, j as usize);
}

fn pop(h: &mut Heap) -> i64 {
    let i = h.i[0];
    h.n -= 1;
    swap(h, 0, h.n);
    down(h, 0);
    i
}

fn down(h: &mut Heap, mut i: i64) {
    loop {
        let left = 2 * i + 1;
        if left >= h.n {
            break;
        }
        let mut j = left;
        let right = left + 1;
        if right < h.n && !less(h, left, right) {
            j = right;
        }
        if less(h, i, j) {
            break;
        }
        swap(h, i, j);
        i = j;
    }
}

fn less(h: &Heap, a: i64, b: i64) -> bool {
    let (i, j) = (h.i[a as usize], h.i[b as usize]);
    h.w[i as usize] < h.w[j as usize]
}

fn update(h: &mut Heap, p: &mut Vec<i64>, i: i64, g: &Matrix) {
    for j in 0..g.n {
        if g.get(i, j) > 0 {
            if h.w[i as usize] + g.get(i, j) < h.w[j as usize] {
                p[j as usize] = i + 1;
                h.w[j as usize] = h.w[i as usize] + g.get(i, j);
                up(h, h.a[j as usize]);
            }
        }
    }
}

fn up(h: &mut Heap, mut j: i64) {
    while j > 0 {
        let i = (j - 1) / 2;
        if i == j || less(h, i, j) {
            break;
        }
        swap(h, i, j);
        j = i;
    }
}

impl Matrix {
    fn get(&self, i: i64, j: i64) -> i64 {
        self.a[(i * self.n + j) as usize]
    }
}
